### ISSUE TITLE
Implement GitHub Checks API

### ISSUE THREAD

**Issue by cjs** [lambci/lambci#106]
The GitHub Checks API (https://developer.github.com/v3/checks/) is a new integration for CI tools.

It's only supported by GitHub Apps, so #105 is a prerequisite.

**Comment by mhart** [comment]
Getting users of LambCI to create a GitHub App per installation is quite an ask – they'd have to enter a whole host of settings that their LambCI instance would need to support: Homepage URL, User authorization callback URL, Webhook URL, etc – all that would require an API Gateway setup, etc. It's a lot more complicated than just an SNS ARN.

Going to leave this open, but don't see it happening without some major work

**Comment by mhart** [comment]
@cjs are you aware of an easy way for users of an open source CI tool to easily setup an app for their own installation?

**Comment by mhart** [comment]
@cjs Like, for example, how do self-hosted Jenkins users use this? Is there an example somewhere of ppl setting this up for their own Jenkins installation?

**Comment by mhart** [comment]
Yeah, that's as I suspected with the App creation. It's a pity there's not a way for open source apps to avoid the whole App creation aspect (ie, just create the App once, for all open source users, and then they just need to enter their own endpoint URL when they install the App).

**Comment by cjs** [comment]
Well, you can, but that means that the creator of the app would incur the costs of deploying and running the gateway (kind of like you do now for the lambci packages). I don't know the details, but I suspect the public Probot examples that people can add to their accounts are using GitHub resources to run on.

**Comment by mhart** [comment]
@cjs is there some good documentation on how to "upgrade" to the checks API for a CI system? Specifically, what events should, and shouldn't, be listened to. And what permissions are required?

Is this right?

<img width="739" alt="screen shot 2018-06-27 at 2 53 35 pm" src="https://user-images.githubusercontent.com/367936/41993570-eb7a0e40-7a19-11e8-84cf-9155518e1b77.png">

<img width="764" alt="screen shot 2018-06-27 at 2 54 59 pm" src="https://user-images.githubusercontent.com/367936/41993650-198a3328-7a1a-11e8-9426-b9b514d8a463.png">

(and all the rest "No access")

**Comment by mhart** [comment]
Actually, it seems that if you have read/write access on Checks, then subscribing to events is unnecessary (I think?)

**Comment by cjs** [comment]
Other than https://developer.github.com/changes/2018-05-07-new-checks-api-public-beta/#integrator-workflows, https://developer.github.com/v3/activity/events/types/#checksuiteevent, https://help.github.com/articles/about-status-checks/#skipping-and-requesting-checks-for-individual-commits and https://developer.github.com/v3/checks/, the GitHub Platform forum is also being used for support (https://platform.github.community/c/integrations).

From my reading of the docs, you just need the `Check suite` and `Check run` events. The read/write on Checks lets you access https://developer.github.com/v3/checks/runs/ but doesn't send you anything.

Let me know if you want help / testing with this.

**Comment by mhart** [comment]
@cjs have been testing this out quite a bit recently. Finally bit the bullet and created a GitHub App to play with, and have been playing with Check Suites and Check Runs extensively.

I've run into an issue (which I also submitted over at the GitHub support form) with Check Suites and Check Runs that you might be able to explain if you have knowledge in this area:

Basically, the event you receive when code is pushed is a Check Suite event.

However, Check Runs can only be created for SHAs, not Check Suites explicitly. There's no way AFAICT to create a Check Run and tell it to attach to a Check Suite – it's just expected that the SHA will determine which Check Suite it gets attached to. Usually this is fine and it will attach to the Check Suite that was created during the code push... except if the SHA already has a Check Suite... say, you reset your branch to an earlier commit and force push.

I've confirmed that I receive Check Suite events in such cases, create a Check Run for the SHA in the Check Suite event, and the Check Run actually returns with a Check Suite ID of the first Check Suite for that SHA, instead of the one the event just fired for. The UI then just displays the most recent Check Suite(s) as Queued... and there's no way to delete them, or attach Check Runs to them, or anything as far as I can tell.

Let me know if you know of any way around this, or how best to deal with it (or if it's a bug)

**Comment by cjs** [comment]
@mhart it looks like Support got back to you with confirmation that you've hit a bug.

**Comment by raman-trantor** [comment]
@cjs @mhart Any update on if this bug was resolved and if we can use lambci with Github Checks API?

**Comment by cjs** [comment]
There was a patch shipped at the end of last July that might have fixed the issue @mhart reported. I haven't been able to manually test to see if that's the case.

**Comment by raman-trantor** [comment]
Thanks for the update @cjs. Definitely will be exciting to have Github Checks API integrated with LambCI.

**Comment by hawkins** [comment]
I've been working with Probot and checks on some projects and would love to see this feature in so we could use lambci instead. Any ideas how I could help out here? Code monkeying or otherwise?
